lib/card_search_query_builder.rb

```diff
@@ -153,100 +153,131 @@ class CardSearchQueryBuilder
     @@term_to_left_join_map = {
     }
 
+    def parse_node(node)
+        key = node.keys[0]
+        case key
+        when :ors
+            children = node[key].kind_of?(Array) ? node[key] : [node[key]]
+            children.map! { |child| parse_node(child) }
+            return '(' + children.join(' OR ') + ')'
+        when :ands
+            children = node[key].kind_of?(Array) ? node[key] : [node[key]]
+            children.map! { |child| parse_node(child) }
+            return '(' + children.join(' AND ') + ')'
+        when :negate
+            child = node[key]
+            return '(NOT ' + parse_node(child) + ')'
+        when :pair
+            pair = node[key]
+            keyword = pair[:keyword].to_s
+            operator = pair[:operator].to_s
+            values = pair[:values].kind_of?(Array) ? pair[:values] : [pair[:values]]
+            values.map! { |v| v[:string].to_s }
+            return parse_pair(keyword, operator, values)
+        when :title
```

Comment:
remind me how title works here.  I would have thought that looking for the default case would land on this, but :title is a proper term already,  no?

Reply:
`:title` is just a string, but it's labelled to give its node a unique identifier. I'm not sure what you're asking, but if it's why this is caught as a specific case rather than dealt with in the default case, it's so that any errors in the grammar get visibly caught.

Reply:
Actually I think I'm going to redo all this using Parslet's Transformer class. I should be able to have a proper go at that on Wednesday.